File: common/random/random.go
package random

import (
	"crypto/rand"
	"fmt"
	"math/big"

	"github.com/pkg/errors"
)

const (
	mustGetRandomIntMaxBits = 5000
)

var (
	zero = big.NewInt(0)
	one  = big.NewInt(1)
	two  = big.NewInt(2)
)

// MustGetRandomInt panics if it is unable to gather entropy from `rand.Reader` or when `bits` is <= 0
func MustGetRandomInt(bits int) *big.Int {
	if bits <= 0 || mustGetRandomIntMaxBits < bits {
		panic(fmt.Errorf("MustGetRandomInt: bits should be positive, non-zero and less than %d", mustGetRandomIntMaxBits))
	}
	// Max random value e.g. 2^256 - 1
	max := new(big.Int)
	max = max.Exp(two, big.NewInt(int64(bits)), nil).Sub(max, one)

	// Generate cryptographically strong pseudo-random int between 0 - max
	n, err := rand.Int(rand.Reader, max)
	if err != nil {
		panic(errors.Wrap(err, "rand.Int failure in MustGetRandomInt!"))
	}
	return n
}

func GetRandomPositiveInt(lessThan *big.Int) *big.Int {
	if lessThan == nil || zero.Cmp(lessThan) != -1 {
		return nil
	}
	var try *big.Int
	for {
		try = MustGetRandomInt(lessThan.BitLen())
		if try.Cmp(lessThan) < 0 && try.Cmp(zero) >= 0 {
			break
		}
	}
	return try
}

func GetRandomPrimeInt(bits int) *big.Int {
	if bits <= 0 {
		return nil
	}
	try, err := rand.Prime(rand.Reader, bits)
	if err != nil ||
		try.Cmp(zero) == 0 {
		// fallback to older method
		for {
			try = MustGetRandomInt(bits)
			if try.ProbablyPrime(50) {
				break
			}
		}
	}
	return try
}

// Generate a random element in the group of all the elements in Z/nZ that
// has a multiplicative inverse.
func GetRandomPositiveRelativelyPrimeInt(n *big.Int) *big.Int {
	if n == nil || zero.Cmp(n) != -1 {
		return nil
	}
	var try *big.Int
	for {
		try = MustGetRandomInt(n.BitLen())
		if IsNumberInMultiplicativeGroup(n, try) {
			break
		}
	}
	return try
}

func IsNumberInMultiplicativeGroup(n, v *big.Int) bool {
	if n == nil || v == nil || zero.Cmp(n) != -1 {
		return false
	}
	gcd := big.NewInt(0)
	return v.Cmp(n) < 0 && v.Cmp(one) >= 0 &&
		gcd.GCD(nil, nil, v, n).Cmp(one) == 0
}

//  Return a random generator of RQn with high probability.  THIS METHOD
//  ONLY WORKS IF N IS THE PRODUCT OF TWO SAFE PRIMES!
// https://github.com/didiercrunch/paillier/blob/d03e8850a8e4c53d04e8016a2ce8762af3278b71/utils.go#L39
func GetRandomGeneratorOfTheQuadraticResidue(n *big.Int) *big.Int {
	r := GetRandomPositiveRelativelyPrimeInt(n)
	return new(big.Int).Mod(new(big.Int).Mul(r, r), n)
}

File: tss-lib-rust/src/common/random.rs
use rand::Rng;
use num_bigint::{BigInt, RandBigInt};
use num_traits::{One, Zero};

const MUST_GET_RANDOM_INT_MAX_BITS: usize = 5000;

pub fn must_get_random_int<R: Rng>(rng: &mut R, bits: usize) -> BigInt {
    if bits <= 0 || bits > MUST_GET_RANDOM_INT_MAX_BITS {
        panic!("MustGetRandomInt: bits should be positive, non-zero and less than {}", MUST_GET_RANDOM_INT_MAX_BITS);
    }
    let max = BigInt::one() << bits;
    rng.gen_bigint_range(&BigInt::zero(), &max)
}

pub fn get_random_positive_int<R: Rng>(rng: &mut R, less_than: &BigInt) -> BigInt {
    if less_than <= &BigInt::zero() {
        return BigInt::zero();
    }
    loop {
        let candidate = must_get_random_int(rng, less_than.bits() as usize);
        if &candidate < less_than {
            return candidate;
        }
    }
}

pub fn get_random_prime_int<R: Rng>(rng: &mut R, bits: usize) -> BigInt {
    loop {
        let candidate = rng.gen_bigint(bits as u64);
        if candidate.is_probably_prime(30) {
            return candidate;
        }
    }
}
#[cfg(test)]
mod tests {
    use super::*;
    use rand::thread_rng;

    #[test]
    fn test_must_get_random_int() {
        let mut rng = thread_rng();
        let random_int = must_get_random_int(&mut rng, 256);
        assert!(random_int.bits() <= 256);
    }

    #[test]
    fn test_get_random_positive_int() {
        let mut rng = thread_rng();
        let less_than = BigInt::from(100);
        let random_int = get_random_positive_int(&mut rng, &less_than);
        assert!(random_int < less_than);
    }
}
